README.rst
RatNav
======

This application is meant to help you drive your car more efficient.

Run it on an embedded Debian system with a camera and speaker attached.

Mount the camera on top of your dashboard so it sees the traffic in front of
you.

Hopefully, the system will accoustically guide you through traffic at a better
efficiency ;)

Source & Issues
===============

Ratnav currently lives here:
    https://github.com/ri0t/ratnav
We use github's nice issue tracker to dissect and kill bugs:
    https://github.com/ri0t/ratnav/issues

Installation
============

Ideally, you're running Debian. You'll need python-opencv and some other stuff
installed:

    sudo apt-get install python-opencv python-numpy

Then you can install ratnav itself:

    sudo python setup.py install

This should be everything necessary to run RatNav.

Copyright, License
==================

This program is free software: you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation, either version 3 of the License, or (at your option) any later
version.

This program is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
PARTICULAR PURPOSE.  See the GNU General Public License for more details.

You should have received a copy of the GNU General Public License along with
this program.  If not, see <http://www.gnu.org/licenses/>.

Contributors
============

This utility was conceived by David Riding and Heiko Weinen at Betahaus in 2014
and is in part based on Robin David's work, which you can find at
https://github.com/RobinDavid/Motion-detection-OpenCV.git

* David Riding
* Heiko 'riot' Weinen <riot@hackerfleet.org>
* Robin David
* KDE4 Granatier application for the 'yippie' sound


debian/changelog
python-ratnav (0.0.1) unstable; urgency=low

    * moved todo items to issue tracker (https://github.com/ri0t/ratnav/issues)
    * state information (so no spamming happens)
    * detecting own movement
    * correctly (hopefully ;) configured application logic
    * added README.rst
    * added gplv3
    * mentions original source
    * added inner movement detection
    * combined delta and contour detection methods in app
    * scrapped recorder
    * cleanups

 -- Heiko 'riot' Weinen <riot@hackerfleet.org> Sun, 10 Aug 2014 13:19:00 +0200


test.py
__author__ = 'riot'
import cv2
import numpy as np
from common import anorm
from matplotlib import pyplot as plt

def filter_matches(kp1, kp2, matches, ratio = 0.75):
    mkp1, mkp2 = [], []
    for m in matches:
        if len(m) == 2 and m[0].distance < m[1].distance * ratio:
            m = m[0]
            mkp1.append( kp1[m.queryIdx] )
            mkp2.append( kp2[m.trainIdx] )
    p1 = np.float32([kp.pt for kp in mkp1])
    p2 = np.float32([kp.pt for kp in mkp2])
    kp_pairs = zip(mkp1, mkp2)
    return p1, p2, kp_pairs

def explore_match(win, img1, img2, kp_pairs, status = None, H = None):
    h1, w1 = img1.shape[:2]
    h2, w2 = img2.shape[:2]
    vis = np.zeros((max(h1, h2), w1+w2), np.uint8)
    vis[:h1, :w1] = img1
    vis[:h2, w1:w1+w2] = img2
    vis = cv2.cvtColor(vis, cv2.COLOR_GRAY2BGR)

    if H is not None:
        corners = np.float32([[0, 0], [w1, 0], [w1, h1], [0, h1]])
        corners = np.int32( cv2.perspectiveTransform(corners.reshape(1, -1, 2), H).reshape(-1, 2) + (w1, 0) )
        cv2.polylines(vis, [corners], True, (255, 255, 255))

    if status is None:
        status = np.ones(len(kp_pairs), np.bool_)
    p1 = np.int32([kpp[0].pt for kpp in kp_pairs])
    p2 = np.int32([kpp[1].pt for kpp in kp_pairs]) + (w1, 0)

    green = (0, 255, 0)
    red = (0, 0, 255)
    white = (255, 255, 255)
    kp_color = (51, 103, 236)
    for (x1, y1), (x2, y2), inlier in zip(p1, p2, status):
        if inlier:
            col = green
            cv2.circle(vis, (x1, y1), 2, col, -1)
            cv2.circle(vis, (x2, y2), 2, col, -1)
        else:
            col = red
            r = 2
            thickness = 3
            cv2.line(vis, (x1-r, y1-r), (x1+r, y1+r), col, thickness)
            cv2.line(vis, (x1-r, y1+r), (x1+r, y1-r), col, thickness)
            cv2.line(vis, (x2-r, y2-r), (x2+r, y2+r), col, thickness)
            cv2.line(vis, (x2-r, y2+r), (x2+r, y2-r), col, thickness)
    vis0 = vis.copy()
    for (x1, y1), (x2, y2), inlier in zip(p1, p2, status):
        if inlier:
            cv2.line(vis, (x1, y1), (x2, y2), green)

    return vis

    # print("Showing")
    # cv2.imshow(win, vis)
    # def onmouse(event, x, y, flags, param):
    #     cur_vis = vis
    #     if flags & cv2.EVENT_FLAG_LBUTTON:
    #         cur_vis = vis0.copy()
    #         r = 8
    #         m = (anorm(p1 - (x, y)) < r) | (anorm(p2 - (x, y)) < r)
    #         idxs = np.where(m)[0]
    #         kp1s, kp2s = [], []
    #         for i in idxs:
    #              (x1, y1), (x2, y2) = p1[i], p2[i]
    #              col = (red, green)[status[i]]
    #              cv2.line(cur_vis, (x1, y1), (x2, y2), col)
    #              kp1, kp2 = kp_pairs[i]
    #              kp1s.append(kp1)
    #              kp2s.append(kp2)
    #         cur_vis = cv2.drawKeypoints(cur_vis, kp1s, flags=4, color=kp_color)
    #         cur_vis[:,w1:] = cv2.drawKeypoints(cur_vis[:,w1:], kp2s, flags=4, color=kp_color)
    #
    #     cv2.imshow(win, cur_vis)
    # cv2.setMouseCallback(win, onmouse)
    # return vis


img = cv2.imread('./images/training/bmw.png', 0)
plate = cv2.imread('./images/query/plate_german.png', 0)

orb = cv2.ORB()

# global thresholding
ret1,th1 = cv2.threshold(img,127,255,cv2.THRESH_BINARY)

# Otsu's thresholding
ret2,th2 = cv2.threshold(img,0,255,cv2.THRESH_BINARY+cv2.THRESH_OTSU)

# Otsu's thresholding after Gaussian filtering
blur = cv2.GaussianBlur(img,(3,3),0)
ret3,th3 = cv2.threshold(blur,0,255,cv2.THRESH_BINARY+cv2.THRESH_OTSU)



# Pick these images as matching sources
img = th2
plate = plate

# Compute keypoints
kp_img, des_img = orb.detectAndCompute(img,None)
kp_plate, des_plate = orb.detectAndCompute(plate,None)

img_kp_plate = cv2.drawKeypoints(plate, kp_plate, color=(0,255,0), flags=0)
img_kp_img = cv2.drawKeypoints(img, kp_img, color=(255 , 0, 0), flags=0)

# plot all the images and their histograms
images = [img, 0, th1,
          img, 0, th2,
          blur, 0, th3,
          img_kp_plate, 0, img_kp_img]
titles = ['Original Noisy Image','Histogram','Global Thresholding (v=127)',
          'Original Noisy Image','Histogram',"Otsu's Thresholding",
          'Gaussian filtered Image','Histogram',"Otsu's Thresholding",
          'Plate KP', 'Histogram', 'Image KP']
plt.figure(0)
for i in xrange(4):
    plt.subplot(5,3,i*3+1),plt.imshow(images[i*3],'gray')
    plt.title(titles[i*3]), plt.xticks([]), plt.yticks([])
    plt.subplot(5,3,i*3+2),plt.hist(images[i*3].ravel(),256)
    plt.title(titles[i*3+1]), plt.xticks([]), plt.yticks([])
    plt.subplot(5,3,i*3+3),plt.imshow(images[i*3+2],'gray')
    plt.title(titles[i*3+2]), plt.xticks([]), plt.yticks([])

plt.show()

# Do the matching

bf = cv2.BFMatcher(cv2.NORM_HAMMING) #, crossCheck=True)

print("Matching")
matches = bf.knnMatch(des_img, trainDescriptors=des_plate, k=2)
print("Filtering")
p1, p2, kp_pairs = filter_matches(kp_img, kp_plate, matches)
print("Displaying")
img_match = explore_match('find_obj', img, plate, kp_pairs)


# Plot the actual matching
plt.figure(1)
plt.imshow(img_match)

plt.show()

ratnav/test.py
__author__ = 'riot'
import cv2
import numpy as np
from common import anorm
from matplotlib import pyplot as plt


def filter_matches(kp1, kp2, matches, ratio=0.75):
    mkp1, mkp2 = [], []
    for m in matches:
        if len(m) == 2 and m[0].distance < m[1].distance * ratio:
            m = m[0]
            mkp1.append(kp1[m.queryIdx])
            mkp2.append(kp2[m.trainIdx])
    p1 = np.float32([kp.pt for kp in mkp1])
    p2 = np.float32([kp.pt for kp in mkp2])
    kp_pairs = zip(mkp1, mkp2)
    return p1, p2, kp_pairs


def explore_match(win, img1, img2, kp_pairs, status=None, H=None):
    h1, w1 = img1.shape[:2]
    h2, w2 = img2.shape[:2]
    vis = np.zeros((max(h1, h2), w1 + w2), np.uint8)
    vis[:h1, :w1] = img1
    vis[:h2, w1:w1 + w2] = img2
    vis = cv2.cvtColor(vis, cv2.COLOR_GRAY2BGR)

    if H is not None:
        corners = np.float32([[0, 0], [w1, 0], [w1, h1], [0, h1]])
        corners = np.int32(cv2.perspectiveTransform(corners.reshape(1, -1, 2), H).reshape(-1, 2) + (w1, 0))
        cv2.polylines(vis, [corners], True, (255, 255, 255))

    if status is None:
        status = np.ones(len(kp_pairs), np.bool_)
    p1 = np.int32([kpp[0].pt for kpp in kp_pairs])
    p2 = np.int32([kpp[1].pt for kpp in kp_pairs]) + (w1, 0)

    green = (0, 255, 0)
    red = (0, 0, 255)
    white = (255, 255, 255)
    kp_color = (51, 103, 236)
    for (x1, y1), (x2, y2), inlier in zip(p1, p2, status):
        if inlier:
            col = green
            cv2.circle(vis, (x1, y1), 2, col, -1)
            cv2.circle(vis, (x2, y2), 2, col, -1)
        else:
            col = red
            r = 2
            thickness = 3
            cv2.line(vis, (x1 - r, y1 - r), (x1 + r, y1 + r), col, thickness)
            cv2.line(vis, (x1 - r, y1 + r), (x1 + r, y1 - r), col, thickness)
            cv2.line(vis, (x2 - r, y2 - r), (x2 + r, y2 + r), col, thickness)
            cv2.line(vis, (x2 - r, y2 + r), (x2 + r, y2 - r), col, thickness)
    vis0 = vis.copy()
    for (x1, y1), (x2, y2), inlier in zip(p1, p2, status):
        if inlier:
            cv2.line(vis, (x1, y1), (x2, y2), green)

    return vis

    # print("Showing")
    # cv2.imshow(win, vis)
    # def onmouse(event, x, y, flags, param):
    #     cur_vis = vis
    #     if flags & cv2.EVENT_FLAG_LBUTTON:
    #         cur_vis = vis0.copy()
    #         r = 8
    #         m = (anorm(p1 - (x, y)) < r) | (anorm(p2 - (x, y)) < r)
    #         idxs = np.where(m)[0]
    #         kp1s, kp2s = [], []
    #         for i in idxs:
    #              (x1, y1), (x2, y2) = p1[i], p2[i]
    #              col = (red, green)[status[i]]
    #              cv2.line(cur_vis, (x1, y1), (x2, y2), col)
    #              kp1, kp2 = kp_pairs[i]
    #              kp1s.append(kp1)
    #              kp2s.append(kp2)
    #         cur_vis = cv2.drawKeypoints(cur_vis, kp1s, flags=4, color=kp_color)
    #         cur_vis[:,w1:] = cv2.drawKeypoints(cur_vis[:,w1:], kp2s, flags=4, color=kp_color)
    #
    #     cv2.imshow(win, cur_vis)
    # cv2.setMouseCallback(win, onmouse)
    # return vis


img = cv2.imread('./images/training/bmw.png', 0)
plate = cv2.imread('./images/query/plate_german.png', 0)

orb = cv2.ORB()

# global thresholding
ret1, th1 = cv2.threshold(img, 127, 255, cv2.THRESH_BINARY)

# Otsu's thresholding
ret2, th2 = cv2.threshold(img, 0, 255, cv2.THRESH_BINARY + cv2.THRESH_OTSU)

# Otsu's thresholding after Gaussian filtering
blur = cv2.GaussianBlur(img, (3, 3), 0)
ret3, th3 = cv2.threshold(blur, 0, 255, cv2.THRESH_BINARY + cv2.THRESH_OTSU)



# Pick these images as matching sources
img = th2
plate = plate

# Compute keypoints
kp_img, des_img = orb.detectAndCompute(img, None)
kp_plate, des_plate = orb.detectAndCompute(plate, None)

img_kp_plate = cv2.drawKeypoints(plate, kp_plate, color=(0, 255, 0), flags=0)
img_kp_img = cv2.drawKeypoints(img, kp_img, color=(255, 0, 0), flags=0)

# plot all the images and their histograms
images = [img, 0, th1,
          img, 0, th2,
          blur, 0, th3,
          img_kp_plate, 0, img_kp_img]
titles = ['Original Noisy Image', 'Histogram', 'Global Thresholding (v=127)',
          'Original Noisy Image', 'Histogram', "Otsu's Thresholding",
          'Gaussian filtered Image', 'Histogram', "Otsu's Thresholding",
          'Plate KP', 'Histogram', 'Image KP']
plt.figure(0)
for i in xrange(4):
    plt.subplot(5, 3, i * 3 + 1), plt.imshow(images[i * 3], 'gray')
    plt.title(titles[i * 3]), plt.xticks([]), plt.yticks([])
    plt.subplot(5, 3, i * 3 + 2), plt.hist(images[i * 3].ravel(), 256)
    plt.title(titles[i * 3 + 1]), plt.xticks([]), plt.yticks([])
    plt.subplot(5, 3, i * 3 + 3), plt.imshow(images[i * 3 + 2], 'gray')
    plt.title(titles[i * 3 + 2]), plt.xticks([]), plt.yticks([])

plt.show()

# Do the matching

bf = cv2.BFMatcher(cv2.NORM_HAMMING)  #, crossCheck=True)

print("Matching")
matches = bf.knnMatch(des_img, trainDescriptors=des_plate, k=2)
print("Filtering")
p1, p2, kp_pairs = filter_matches(kp_img, kp_plate, matches)
print("Displaying")
img_match = explore_match('find_obj', img, plate, kp_pairs)


# Plot the actual matching
plt.figure(1)
plt.imshow(img_match)

plt.show()

ratnav/app.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-

# RatNav
# ======================================
# Copyright (C) 2014 RatNav Contributors
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.


import time
import pygame

import cv2.cv as cv
import numpy as np

try:
    import picamera
    import io

    PICAM = True
except ImportError:
    PICAM = False


# Audio files
MOVE = './audio/move.wav'  # A call to drive
MOVING = './audio/moving.wav'  # When we're really driving again
STANDING = './audio/standing.wav'  # When we had to stop (this one might annoy)


def log(*what):
    """Simple unfancy logging"""

    msg = ""
    for item in what:
        msg += str(item)

    print(msg)


class RatNavApp():
    """
    The RatNav Application object

    Just instantiate and run, e.g. like this:
    > RatNav = RatNavApp(do_audio=False, show_windows=False)
    > RatNav.run()

    This would give you a ratnav without display and audio output.
    Which is still useful for debugging purposes, sometimes.

    RatNav will observe your first installed camera for movement
    changes. If it detects that the camera is standing still (no
    detectable movement happens), it switches into waiting mode
    and observes only the center of your viewport.
    If it detects a movement here while waiting, it will alert
    you to this change.
    Once it detects that you're driving again, it will play a
    cheering alert and wait until movement stops, which will
    be alerted, too.
    """

    def on_threshold_change(self, val):
        """Called when threshold slider is changed"""

        self.threshold = val

    def capture_frame(self):

        if PICAM:
            with picamera.PiCamera() as camera:
                camera.capture('snapshot.jpg')
                image = cv.LoadImage('snapshot.jpg')
        else:
            image = cv.QueryFrame(self.capture)  # Take a frame to init recorder

        return image


    def __init__(self, threshold=25, show_windows=True, do_audio=True):
        """Set up cv buffers, audiofiles and states"""

        self.font = None
        self.mode = 'contours'
        self.show = show_windows  # Either or not show the 2 windows
        self.frame = None

        if PICAM:
            log("Using RPi camera")
            #self.capture = picamera.PiCamera()
            #self.capture.start_preview()
            #self.stream = io.BytesIO()
        else:
            log("Using normal cv camera")
            self.capture = cv.CaptureFromCAM(0)
        log("Camera opened")
        self.frame = self.capture_frame()  # Take a frame to init buffer sizes
        log("First frame taken: ", type(self.frame), self.frame)

        self.gray_frame = cv.CreateImage(cv.GetSize(self.frame), cv.IPL_DEPTH_8U, 1)
        self.average_frame = cv.CreateImage(cv.GetSize(self.frame), cv.IPL_DEPTH_32F, 3)
        self.absdiff_frame = None
        self.previous_frame = None

        # Will hold the thresholded result
        self.res = cv.CreateMat(self.frame.height, self.frame.width, cv.CV_8U)

        self.frame1gray = cv.CreateMat(self.frame.height, self.frame.width, cv.CV_8U)  # Gray frame at t-1
        cv.CvtColor(self.frame, self.frame1gray, cv.CV_RGB2GRAY)

        self.frame2gray = cv.CreateMat(self.frame.height, self.frame.width, cv.CV_8U)  # Gray frame at t

        self.width = self.frame.width
        self.height = self.frame.height
        self.nb_pixels = self.width * self.height

        self.inner = (self.width / 4, self.height / 4, (self.width / 2), (self.height / 2))
        log("Input format: ", self.width, 'x', self.height)
        log("Scanning for car movement in: ", self.inner)

        self.surface = self.width * self.height
        self.currentsurface = 0
        self.currentcontours = None
        self.threshold = threshold
        #self.trigger_time = 0   # Hold timestamp of the last detection
        self.move_time = 0
        self.alert_time = 0

        self.standing = True
        self.moving = False

        self.do_audio = do_audio

        self.sounds = {}

        if do_audio:
            pygame.init()

            self.sounds[MOVING] = pygame.mixer.Sound(MOVING)
            self.sounds[MOVE] = pygame.mixer.Sound(MOVE)
            self.sounds[STANDING] = pygame.mixer.Sound(STANDING)

        if show_windows:
            cv.NamedWindow("Image")
            cv.CreateTrackbar("Detection treshold: ", "Image", self.threshold, 100, self.on_threshold_change)

    def run(self):
        """The actual main loop.
        * Queries a new camera picture
        * Processes the image
        * Checks for movement

        """

        started = time.time()

        while True:

            currentframe = self.capture_frame()
            instant = time.time()  # Get timestamp o the frame

            self.process_image(currentframe)  # Process the image

            if self.has_movement():
                if self.standing:  # we _we're_ standing
                    log("We detected a first movement.")
                    self.standing = False
                    self.move_time = instant
                if not self.moving and instant > self.move_time + 3:  # once we're sure, alert
                    log("We are moving.")
                    self.alert(MOVING)
                    self.moving = True

                    #if instant > started + 10:   # Wait 5 second after the webcam start for luminosity adjusting etc..
            else:
                if self.moving:  # seems like we're standing again
                    log("We stand still.")
                    self.alert(STANDING)
                    self.moving = False  # because we didn't see movement for at least a frame
                self.standing = True
                if self.has_movement_in_rect(self.currentcontours, self.inner):
                    log("We should move.")
                    self.alert(MOVE)

            if self.mode == 'contours':
                cv.DrawContours(currentframe, self.currentcontours, (0, 0, 255), (0, 255, 0), 1, 2, cv.CV_FILLED)

            if self.show:
                cv.ShowImage("Image", currentframe)

            c = cv.WaitKey(1) % 0x100
            if c == 27 or c == 10:  # Break if user enters 'Esc'.
                break

    def process_image(self, frame):
        """Decides which processing to use. No, not exactly a factory."""

        if self.mode == 'contours':
            self.process_for_contours(frame)
        else:
            self.process_for_threshold(frame)

    def process_for_contours(self, curframe):
        """Do cv contour processing."""

        cv.Smooth(curframe, curframe)  # Remove false positives

        if not self.absdiff_frame:  # For the first time put values in difference, temp and moving_average
            self.absdiff_frame = cv.CloneImage(curframe)
            self.previous_frame = cv.CloneImage(curframe)
            cv.Convert(curframe, self.average_frame)  # Should convert because after runningavg take 32F pictures
        else:
            cv.RunningAvg(curframe, self.average_frame, 0.05)  # Compute the average

        cv.Convert(self.average_frame, self.previous_frame)  # Convert back to 8U frame

        cv.AbsDiff(curframe, self.previous_frame, self.absdiff_frame)  # moving_average - curframe

        cv.CvtColor(self.absdiff_frame, self.gray_frame, cv.CV_RGB2GRAY)  # Convert to gray otherwise can't do threshold
        cv.Threshold(self.gray_frame, self.gray_frame, 50, 255, cv.CV_THRESH_BINARY)

        cv.Dilate(self.gray_frame, self.gray_frame, None, 15)  # to get object blobs
        cv.Erode(self.gray_frame, self.gray_frame, None, 10)

    def process_for_threshold(self, frame):
        """Do cv threshold processing."""

        cv.CvtColor(frame, self.frame2gray, cv.CV_RGB2GRAY)

        #Absdiff to get the difference between to the frames
        cv.AbsDiff(self.frame1gray, self.frame2gray, self.res)

        #Remove the noise and do the threshold
        cv.Smooth(self.res, self.res, cv.CV_BLUR, 5, 5)
        cv.MorphologyEx(self.res, self.res, None, None, cv.CV_MOP_OPEN)
        cv.MorphologyEx(self.res, self.res, None, None, cv.CV_MOP_CLOSE)
        cv.Threshold(self.res, self.res, 10, 255, cv.CV_THRESH_BINARY_INV)

    def has_movement_in_rect(self, contours, rect):
        """Checks for movement only in the given rectangle."""

        result = False

        for contour in contours:
            if type(contour) == tuple:
                crect = cv.BoundingRect([contour])
            else:
                crect = cv.BoundingRect(contour)
            cx, cy, cw, ch = crect
            x, y, w, h = rect
            if (cx > x) and (cx < x + w):
                if (cy > y) and (cy < y + h):
                    result = True

        return result

    def has_movement(self):
        """
        Tests for movement in the image.
        Currently also tests for movement in the inner rectangle.
        """

        result = False
        if self.mode == 'contours':
            #  Find contours

            storage = cv.CreateMemStorage(0)
            contours = cv.FindContours(self.gray_frame, storage, cv.CV_RETR_EXTERNAL, cv.CV_CHAIN_APPROX_SIMPLE)

            self.currentcontours = contours  # Save contours

            while contours:  # For all contours compute the area
                self.currentsurface += cv.ContourArea(contours)
                contours = contours.h_next()

            avg = (self.currentsurface * 100) / self.surface  # Calculate the average of contour area on the total size
            self.currentsurface = 0  # Put back the current surface to 0

            if avg > self.threshold:
                result = True
        else:
            nb = 0  # Will hold the number of black pixels

            for x in range(self.height):  # Iterate the hole image
                for y in range(self.width):
                    if self.res[x, y] == 0.0:  # If the pixel is black keep it
                        nb += 1
            avg = (nb * 100.0) / self.nb_pixels  # Calculate the average of black pixel in the image

            if avg > self.threshold:  # If over the ceiling trigger the alarm
                result = True

        return result

    def alert(self, soundname):
        """
        Should play the audio file upon traffic movement alert.
        """

        if self.alert_time + 5 < time.time():
            if self.do_audio:
                log("Playing audio '%s'" % soundname)
                self.sounds[soundname].play()

            self.alert_time = time.time()
        else:
            log("Not alerting.")


ratnav/logger.py
#!/usr/bin/env python3
# -*- coding: UTF-8 -*-

import time
import sys

debug = 10
info = 20
warn = 30
error = 40
critical = 50
off = 100

lvldata = {10: ['DEBUG', '\033[1;97m'],
           20: ['INFO', '\033[1;92m'],
           30: ['WARN', '\033[1;94m'],
           40: ['ERROR', '\033[1;91m'],
           50: ['CRITICAL', '\033[1;95m'],
}

count = 0

logfile = "/var/log/ratnav/service.log"
verbosity = {'global': debug,
             'file': off,
             'console': debug}

start = time.time()


def log(*what, **kwargs):
    if 'lvl' in kwargs:
        lvl = kwargs['lvl']
        if lvl < verbosity['global']:
            return
    else:
        lvl = info

    global count
    global start
    count += 1

    now = time.time() - start
    msg = "[%s] : %5s : %.5f : %5i :" % (time.asctime(),
                                         lvldata[lvl][0],
                                         now,
                                         count)

    for thing in what:
        msg += " "
        msg += str(thing)

    if lvl >= verbosity['file']:
        try:
            f = open(logfile, "a")
            f.write(msg)
            f.flush()
            f.close()
        except IOError:
            print("Can't open logfile for writing!")
            sys.exit(23)

    if lvl >= verbosity['console']:
        print(lvldata[lvl][1], str(msg), '\033[0m')

ratnav/find_obj.py
#!/usr/bin/env python

'''
Feature-based image matching sample.

USAGE
  find_obj.py [--feature=<sift|surf|orb>[-flann]] [ <image1> <image2> ]

  --feature  - Feature to use. Can be sift, surf of orb. Append '-flann' to feature name
                to use Flann-based matcher instead bruteforce.

  Press left mouse button on a feature point to see its matching point.
'''

import numpy as np
import cv2
from common import anorm, getsize

FLANN_INDEX_KDTREE = 1  # bug: flann enums are missing
FLANN_INDEX_LSH = 6


def init_feature(name):
    chunks = name.split('-')
    if chunks[0] == 'sift':
        detector = cv2.SIFT()
        norm = cv2.NORM_L2
    elif chunks[0] == 'surf':
        detector = cv2.SURF(800)
        norm = cv2.NORM_L2
    elif chunks[0] == 'orb':
        detector = cv2.ORB(400)
        norm = cv2.NORM_HAMMING
    else:
        return None, None
    if 'flann' in chunks:
        if norm == cv2.NORM_L2:
            flann_params = dict(algorithm=FLANN_INDEX_KDTREE, trees=5)
        else:
            flann_params = dict(algorithm=FLANN_INDEX_LSH,
                                table_number=6,  # 12
                                key_size=12,  # 20
                                multi_probe_level=1)  #2
        matcher = cv2.FlannBasedMatcher(flann_params, {})  # bug : need to pass empty dict (#1329)
    else:
        matcher = cv2.BFMatcher(norm)
    return detector, matcher


def filter_matches(kp1, kp2, matches, ratio=0.75):
    mkp1, mkp2 = [], []
    for m in matches:
        if len(m) == 2 and m[0].distance < m[1].distance * ratio:
            m = m[0]
            mkp1.append(kp1[m.queryIdx])
            mkp2.append(kp2[m.trainIdx])
    p1 = np.float32([kp.pt for kp in mkp1])
    p2 = np.float32([kp.pt for kp in mkp2])
    kp_pairs = zip(mkp1, mkp2)
    return p1, p2, kp_pairs


def explore_match(win, img1, img2, kp_pairs, status=None, H=None):
    h1, w1 = img1.shape[:2]
    h2, w2 = img2.shape[:2]
    vis = np.zeros((max(h1, h2), w1 + w2), np.uint8)
    vis[:h1, :w1] = img1
    vis[:h2, w1:w1 + w2] = img2
    vis = cv2.cvtColor(vis, cv2.COLOR_GRAY2BGR)

    if H is not None:
        corners = np.float32([[0, 0], [w1, 0], [w1, h1], [0, h1]])
        corners = np.int32(cv2.perspectiveTransform(corners.reshape(1, -1, 2), H).reshape(-1, 2) + (w1, 0))
        cv2.polylines(vis, [corners], True, (255, 255, 255))

    if status is None:
        status = np.ones(len(kp_pairs), np.bool_)
    p1 = np.int32([kpp[0].pt for kpp in kp_pairs])
    p2 = np.int32([kpp[1].pt for kpp in kp_pairs]) + (w1, 0)

    green = (0, 255, 0)
    red = (0, 0, 255)
    white = (255, 255, 255)
    kp_color = (51, 103, 236)
    for (x1, y1), (x2, y2), inlier in zip(p1, p2, status):
        if inlier:
            col = green
            cv2.circle(vis, (x1, y1), 2, col, -1)
            cv2.circle(vis, (x2, y2), 2, col, -1)
        else:
            col = red
            r = 2
            thickness = 3
            cv2.line(vis, (x1 - r, y1 - r), (x1 + r, y1 + r), col, thickness)
            cv2.line(vis, (x1 - r, y1 + r), (x1 + r, y1 - r), col, thickness)
            cv2.line(vis, (x2 - r, y2 - r), (x2 + r, y2 + r), col, thickness)
            cv2.line(vis, (x2 - r, y2 + r), (x2 + r, y2 - r), col, thickness)
    vis0 = vis.copy()
    for (x1, y1), (x2, y2), inlier in zip(p1, p2, status):
        if inlier:
            cv2.line(vis, (x1, y1), (x2, y2), green)

    cv2.imshow(win, vis)

    def onmouse(event, x, y, flags, param):
        cur_vis = vis
        if flags & cv2.EVENT_FLAG_LBUTTON:
            cur_vis = vis0.copy()
            r = 8
            m = (anorm(p1 - (x, y)) < r) | (anorm(p2 - (x, y)) < r)
            idxs = np.where(m)[0]
            kp1s, kp2s = [], []
            for i in idxs:
                (x1, y1), (x2, y2) = p1[i], p2[i]
                col = (red, green)[status[i]]
                cv2.line(cur_vis, (x1, y1), (x2, y2), col)
                kp1, kp2 = kp_pairs[i]
                kp1s.append(kp1)
                kp2s.append(kp2)
            cur_vis = cv2.drawKeypoints(cur_vis, kp1s, flags=4, color=kp_color)
            cur_vis[:, w1:] = cv2.drawKeypoints(cur_vis[:, w1:], kp2s, flags=4, color=kp_color)

        cv2.imshow(win, cur_vis)

    cv2.setMouseCallback(win, onmouse)
    return vis


if __name__ == '__main__':
    print __doc__

    import sys, getopt

    opts, args = getopt.getopt(sys.argv[1:], '', ['feature='])
    opts = dict(opts)
    feature_name = opts.get('--feature', 'sift')
    try:
        fn1, fn2 = args
    except:
        fn1 = '../c/box.png'
        fn2 = '../c/box_in_scene.png'

    img1 = cv2.imread(fn1, 0)
    img2 = cv2.imread(fn2, 0)
    detector, matcher = init_feature(feature_name)
    if detector != None:
        print 'using', feature_name
    else:
        print 'unknown feature:', feature_name
        sys.exit(1)

    kp1, desc1 = detector.detectAndCompute(img1, None)
    kp2, desc2 = detector.detectAndCompute(img2, None)
    print 'img1 - %d features, img2 - %d features' % (len(kp1), len(kp2))

    def match_and_draw(win):
        print 'matching...'
        raw_matches = matcher.knnMatch(desc1, trainDescriptors=desc2, k=2)  #2
        p1, p2, kp_pairs = filter_matches(kp1, kp2, raw_matches)
        if len(p1) >= 4:
            H, status = cv2.findHomography(p1, p2, cv2.RANSAC, 5.0)
            print '%d / %d  inliers/matched' % (np.sum(status), len(status))
        else:
            H, status = None, None
            print '%d matches found, not enough for homography estimation' % len(p1)

        vis = explore_match(win, img1, img2, kp_pairs, status, H)

    match_and_draw('find_obj')
    cv2.waitKey()
    cv2.destroyAllWindows()
